rfcs/0013-configurable-and-composable-storage.md

```diff
@@ -0,0 +1,600 @@
+    //--
+    vec_of_refs: StorageVec<StorageRefBox<StVecOfU64>> := [
+        StorageRef::Null,
+        storage.vec_1.as_ref(), // <<-- Note using the `storage` keyword here, as well as `as_ref`.
+        storage.pair_1.first().as_ref(),
+        storage.map_1.get("abc").as_ref(),
+        StorageRef::Ref(some_const_fn_that_returns_the_storage_key_of_the_referenced_storage()),
+    ],
+}
+```
+
+For more examples on using storage references, see the [`demo_contract_storage_references.sw`](../files/0013-configurable-and-composable-storage/contracts/demo_contract_storage_references.sw).
+
+## Breaking changes
+
+[breaking-changes]: #breaking-changes
+
+This proposal brings breaking changes in:
+- `storage` declarations.
+- existing STD storage types like, e.g., `StorageVec`.
+- low-level storage API.
+- the code that deals with the storage, via any of the above approaches.
+
+Additionally, values stored using the current storage API cannot always be read with the new API and vice versa, because some of the changes will result in a different layout of the stored values within the storage.
+
+To support explaining breaking changes to existing users and to ease migration we will:
+- utilize [expressive diagnostics](./0012-expressive-diagnostics.md) to provide detailed explanations on how to fix compiler errors and migrate the code.
+- where applicable, provide scripts to semi-automate the migration by converting the existing code to the new syntax and semantics.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation of the proposed features relies on a couple of other language features that need to be implemented first. We will only list them here, since their detailed explanation, in certain cases, could require a separate RFC. To follow the reference-level explanation of the configurable and composable storage, it is sufficient to have an intuitive understanding of these language features. Where needed, they are explained in enough detail in the [sample code](../files/0013-configurable-and-composable-storage/).
+
+The [sample code](../files/0013-configurable-and-composable-storage/) provides a detailed view on how the Sway STD part of this RFC will be implemented and used.
+
+Here, we will focus on the compiler support for the feature.
+
+The additional language features that are the prerequisite to implement this RFC are:
+- const functions and traits.
+- typed slices.
+- marker traits.
+- negative impls.
+- type aliases with generic parameters and trait constraints.
+
+## The `Storage` trait
+
+[the-storage-trait]: #the-storage-trait
+
+The `Storage` trait is [defined](../files/0013-configurable-and-composable-storage/sway-libs/storage.sw) as:
+
+```Sway
+pub trait Storage {
+    type Value;
+    type Config;
+
+    const fn new(self_key: &StorageKey) -> Self;
+
+    const fn internal_get_config(self_key: &StorageKey, value: &Self::Value) -> Self::Config;
```

Comment:
If we go with `StorageInit` instead of `Config`, maybe this can be `internal_make_initializer`

Reply:
I like the verb `make`! I didn't like `get` and was thinking about alternatives, but didn't like them, e.g. `create`, either. They all sounded very "runtimeysh". `make` somehow does not have that apeal. Here is your self key and the value to store, _make_ me the <config (good name still to find :smile:)>.

As mentioned, `initializer` can mislead and in the and, the final name behind `make` will depend on how we name the `Config` associated type.